File: dotnet-7/EmitLogDirect/EmitLogDirect.cs
using RabbitMQ.Client;
using System.Text;

var factory = new ConnectionFactory() { HostName = "localhost" };
using var connection = factory.CreateConnection();
using var channel = connection.CreateModel();

channel.ExchangeDeclare(exchange: "direct_logs", type: "direct");

var severity = (args.Length > 0) ? args[0] : "info";
var message = (args.Length > 1) ? string.Join(" ", args.Skip(1).ToArray()) : "Hello World!";
var body = Encoding.UTF8.GetBytes(message);
channel.BasicPublish(exchange: "direct_logs", routingKey: severity, basicProperties: null, body: body);
Console.WriteLine($" [x] Sent '{severity}':'{message}'");

Console.WriteLine(" Press [enter] to exit.");
Console.ReadLine();

File: php-thesis/routing/emit_log_direct.php
<?php

declare(strict_types=1);

use Thesis\Amqp\Client;
use Thesis\Amqp\Config;
use Thesis\Amqp\Message;

require_once __DIR__.'/../vendor/autoload.php';

$client = new Client(
    new Config(
        urls: ['rabbitmq:5672'],
    ),
);

$channel = $client->channel();

$channel->exchangeDeclare(exchange: 'direct_logs', exchangeType: 'direct');

$severity = $argv[1] ?? 'info';

$channel->publish(
    new Message(
        body: $data = implode(' ', array_slice($argv, 2)) ?: 'Hello World!',
    ),
    exchange: 'direct_logs',
    routingKey: $severity,
);

echo " [x] Sent {$severity}: {$data} \n";

$client->disconnect();
